File: fingerprint-auth-demo-backend/src/main/java/com/vunm/demo/domain/service/FingerprintApiService.java
package com.vunm.demo.domain.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.vunm.demo.config.FingerprintConfig;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.HashMap;
import java.util.Map;

@Slf4j
@Service
@RequiredArgsConstructor
public class FingerprintApiService {
    private final RestTemplate restTemplate;
    private final FingerprintConfig fingerprintConfig;
    private final ObjectMapper objectMapper;

    /**
     * Get visitor information from FingerprintJS Pro API
     * 
     * @param visitorId The visitor ID from the client
     * @return Map containing visitor information including bot detection data
     */
    public Map<String, Object> getVisitorInfo(String visitorId) {
        if (!fingerprintConfig.isConfigured()) {
            log.warn("FingerprintJS Pro API key not configured, skipping API call");
            return new HashMap<>();
        }

        try {
            String url = UriComponentsBuilder
                .fromHttpUrl(fingerprintConfig.getApiUrl())
                .path("/visitors/" + visitorId)
                .queryParam("region", fingerprintConfig.getApiRegion())
                .build()
                .toUriString();

            HttpHeaders headers = new HttpHeaders();
            headers.set("Auth-API-Key", fingerprintConfig.getApiKey());
            headers.set("Content-Type", "application/json");

            HttpEntity<String> entity = new HttpEntity<>(headers);
            
            log.info("Calling FingerprintJS Pro API for visitor: {}", visitorId);
            ResponseEntity<String> response = restTemplate.exchange(
                url, 
                HttpMethod.GET, 
                entity, 
                String.class
            );

            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                JsonNode rootNode = objectMapper.readTree(response.getBody());
                Map<String, Object> result = new HashMap<>();
                
                // Extract bot detection data if available
                if (rootNode.has("products") && rootNode.get("products").has("botd")) {
                    JsonNode botdNode = rootNode.get("products").get("botd");
                    if (botdNode.has("data") && botdNode.get("data").has("bot")) {
                        JsonNode botNode = botdNode.get("data").get("bot");
                        
                        double probability = botNode.has("probability") ? 
                            botNode.get("probability").asDouble() : 0.0;
                        
                        String type = botNode.has("type") ? 
                            botNode.get("type").asText() : "unknown";
                        
                        result.put("botProbability", probability);
                        result.put("botType", type);
                        result.put("isBot", probability > 0.8); // Consider as bot if probability > 80%
                        
                        log.info("Bot detection for visitor {}: probability={}, type={}", 
                            visitorId, probability, type);
                    }
                }
                
                // Add raw response for debugging
                result.put("rawResponse", response.getBody());
                
                return result;
            } else {
                log.error("Failed to get visitor info from FingerprintJS Pro API: {}", response.getStatusCode());
                return new HashMap<>();
            }
        } catch (Exception e) {
            log.error("Error calling FingerprintJS Pro API: {}", e.getMessage(), e);
            return new HashMap<>();
        }
    }
}

File: fingerprint-auth-demo-backend-go/services/fingerprint_api_service.go
package services

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"

	"github.com/vunm/fingerprint-auth-demo-backend-go/config"
)

// FingerprintAPIResponse represents the response from the FingerprintJS Pro API
type FingerprintAPIResponse struct {
	Products struct {
		Botd struct {
			Data struct {
				Bot struct {
					Probability float64 `json:"probability"`
					Type        string  `json:"type"`
				} `json:"bot"`
			} `json:"data"`
		} `json:"botd"`
	} `json:"products"`
}

// GetVisitorInfo gets visitor information from the FingerprintJS Pro API
func GetVisitorInfo(visitorID string) (map[string]interface{}, error) {
	result := make(map[string]interface{})

	// Check if API key is configured
	if config.AppConfig.FingerprintApiKey == "" {
		log.Println("FingerprintJS Pro API key not configured, skipping API call")
		return result, nil
	}

	// Build the API URL
	url := fmt.Sprintf("%s/visitors/%s?region=%s", 
		config.AppConfig.FingerprintApiUrl, 
		visitorID, 
		config.AppConfig.FingerprintApiRegion)

	// Create the request
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("error creating request: %w", err)
	}

	// Add headers
	req.Header.Set("Auth-API-Key", config.AppConfig.FingerprintApiKey)
	req.Header.Set("Content-Type", "application/json")

	// Send the request
	client := &http.Client{}
	log.Printf("Calling FingerprintJS Pro API for visitor: %s", visitorID)
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error calling FingerprintJS Pro API: %w", err)
	}
	defer resp.Body.Close()

	// Check response status
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("FingerprintJS Pro API returned status: %d", resp.StatusCode)
	}

	// Read response body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading response body: %w", err)
	}

	// Parse response
	var apiResponse FingerprintAPIResponse
	err = json.Unmarshal(body, &apiResponse)
	if err != nil {
		return nil, fmt.Errorf("error parsing response: %w", err)
	}

	// Extract bot detection data
	botProbability := apiResponse.Products.Botd.Data.Bot.Probability
	botType := apiResponse.Products.Botd.Data.Bot.Type
	isBot := botProbability > 0.8 // Consider as bot if probability > 80%

	result["botProbability"] = botProbability
	result["botType"] = botType
	result["isBot"] = isBot
	result["rawResponse"] = string(body)

	log.Printf("Bot detection for visitor %s: probability=%f, type=%s", 
		visitorID, botProbability, botType)

	return result, nil
}